File: agent-service/app/schemas/npc.py
from typing import Literal
from pydantic import BaseModel, Field, ConfigDict

ActionVerb = Literal[
    "MOVE", "TALK", "ATTACK", "DEFEND", "FLEE", "BUY", "SELL", "TRADE", "USE", "WAIT"
]


class Coordinate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    x: int = Field(alias="X")
    y: int = Field(alias="Y")


class WeaponInfo(BaseModel):
    name: str
    power: int
    durability: int
    max_durability: int


class MerchantInfo(BaseModel):
    name: str
    stock: list[str] | None = None


class AdjacentMonsterAlert(BaseModel):
    species_or_name: str
    in_safe_zone: bool


class NegotiationAlert(BaseModel):
    partner: str
    turns: int


class PendingReplyAlert(BaseModel):
    model_config = ConfigDict(populate_by_name=True)
    from_whom: str = Field(alias="from")
    text: str


class PerceptionAlerts(BaseModel):
    adjacent_monster: AdjacentMonsterAlert | None = None
    negotiation_stalled: NegotiationAlert | None = None
    pending_reply: PendingReplyAlert | None = None


class Perception(BaseModel):
    name: str
    role: str
    personality: str
    faction: str
    current_mood: str
    background: str
    objective: str
    current_goal: str | None = ""
    relationships: list[str] | None = None
    likes: list[str] | None = None
    dislikes: list[str] | None = None
    known_secrets: list[str] | None = None
    level: int
    hp: int
    max_hp: int
    base_attack: int
    weapon_info: WeaponInfo | None = None
    combat_style: str | None = None
    skills: list[str] | None = None
    gold: int
    inventory_names: list[str] | None = None
    nearby_merchant: MerchantInfo | None = None
    position: Coordinate
    biome: str
    tile_description: str
    is_safe_zone: bool
    nearby_pois: list[str] | None = None
    trust_fear_lines: list[str] | None = None
    visible_entities: list[str] | None = None
    relevant_memories: list[str] | None = None
    alerts: PerceptionAlerts | None = None


class AgentAction(BaseModel):
    internal_thought: str
    action: ActionVerb
    target_coordinates: list[int] | None = None
    target: str | None = None
    dialogue: str | None = None
    skill: str | None = None
    goal_coordinates: list[int] | None = None
    item: str | None = None
    gold: int | None = None
    trust_shift: int | None = None


class GroupDecideRequest(BaseModel):
    members: list[Perception]


class GroupDecideResponse(BaseModel):
    actions: dict[str, AgentAction]

File: engine/models/npc.go
package models

type Position struct {
	X int
	Y int
}

type Item struct {
	Name          string
	Power         int
	Durability    int
	MaxDurability int

	Kind string

	HealAmount int

	Price int
}

type Tile struct {
	Biome       string
	Description string
	IsSafeZone  bool
}

type POI struct {
	Name string `json:"name"`
	Icon string `json:"icon"`
	X    int    `json:"x"`
	Y    int    `json:"y"`
}

type NPC struct {
	Name         string
	Role         string
	Background   string
	Objective    string
	Position     Position
	VisionRadius int
	Memory       []string
	LastAction   string

	Personality   string
	Relationships []string
	Faction       string
	CombatStyle   string

	Likes    []string
	Dislikes []string

	KnownSecrets []string

	CurrentMood string
	CurrentGoal string

	Trust map[string]int
	Fear  map[string]int

	Skills []string

	HP         int
	MaxHP      int
	Level      int
	XP         int
	BaseAttack int
	Equipped   *Item

	IsLLM              bool
	IsPlayerControlled bool

	Species string

	GoalPosition *Position

	LastSignature   string
	TicksSinceThink int
	LastProvider    string

	HeartbeatTicks int

	Defending bool

	Gold int

	Inventory []*Item

	IsMerchant bool
	Shop       []Item

	TicksSinceLastHit int

	BlockedTicks int

	PrevPosition *Position

	NegotiationPartner string
	NegotiationTicks   int

	LastPrintedLine string
	IdleRepeatCount int

	PendingReplyFrom    string
	PendingReplyText    string
	PendingReplySetTurn int

	LastAttackTarget string
}

type AgentAction struct {
	InternalThought   string     `json:"internal_thought"`
	Action            ActionVerb `json:"action"`
	TargetCoordinates []int      `json:"target_coordinates"`
	Target            string     `json:"target"`
	Dialogue          string     `json:"dialogue"`

	Skill string `json:"skill,omitempty"`

	GoalCoordinates []int `json:"goal_coordinates,omitempty"`

	Item string `json:"item,omitempty"`

	Gold int `json:"gold,omitempty"`

	TrustShift int `json:"trust_shift,omitempty"`
}
